### PR TITLE
Update SMP v2 notebooks to use latest PyTorch 2.3.1, TSM 2.5.0 release

### PR DESCRIPTION
**Pull request by viclzhu**
*Issue #, if available:*

*Description of changes:*

*Testing done:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

- [x] I have verified that my PR does not contain any new notebook/s which demonstrate a SageMaker functionality already showcased by another existing notebook in the repository
- [x] I have read the [CONTRIBUTING](https://github.com/aws/amazon-sagemaker-examples/blob/default/CONTRIBUTING.md) doc and adhered to the guidelines regarding folder placement, notebook naming convention and example notebook best practices
- [x] I have updated the necessary documentation, including the README of the appropriate folder as well as the index.rst file
- [x] I have tested my notebook(s) and ensured it runs end-to-end
- [x] I have linted my notebook(s) and code using `black-nb -l 100 {path}/{notebook-name}.ipynb`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: training/distributed_training/pytorch/model_parallel_v2/shared-scripts/train_lib.py [hunk 1]
```diff
@@ -110,6 +119,30 @@ def reduce_loss(loss):
     loss_scalar = loss_detached.item() / dist.get_world_size()
     return loss_scalar
 
+def get_batch_for_cp_rank(batch):
+    # Based on https://tiny.amazon.com/1bcmbuhje/githNVIDNeMoblob58d6nemocoll
```
Comment (SuhitK) [inline training/distributed_training/pytorch/model_parallel_v2/shared-scripts/train_lib.py:123]: Replace this amazon based tiny url with generic url

File: training/distributed_training/pytorch/model_parallel_v2/shared-scripts/train_lib.py [hunk 2]
```diff
@@ -110,6 +119,30 @@ def reduce_loss(loss):
...
+    cp_size = tsm.state.cp_size
+    cp_rank = tsm.state.cp_rank
+    if cp_size > 1:
+        return_batch = []
+        for val in batch:
+            if val is not None:
+                seq_dim = 1  # TODO(skodgule): In train_lib.py check seq dims of attn mask, here its assumed 1
```
Comment (SuhitK) [inline training/distributed_training/pytorch/model_parallel_v2/shared-scripts/train_lib.py:130]: Remove the TODO

File: training/distributed_training/pytorch/model_parallel_v2/shared-scripts/train_lib.py [hunk 3]
```diff
@@ -322,6 +377,8 @@ def train(
                 if args.preserve_np_state > 0:
                     np.random.set_state(cur_state)
 
+            # integration test
```
Comment (SuhitK) [inline training/distributed_training/pytorch/model_parallel_v2/shared-scripts/train_lib.py:380]: Nit: Remove comment

File: training/distributed_training/pytorch/model_parallel_v2/shared-scripts/train_utils.py [hunk 4]
```diff
@@ -192,7 +192,7 @@ def new_attn(
                 attn_weights,
             )
 
-        if args.model_type == "llama_v2":
+        if args.model_type == "llama_v2" or args.model_type == "llama_v3":
             # pre 4.34 we use rubik's class
```
Comment (SuhitK) [inline training/distributed_training/pytorch/model_parallel_v2/shared-scripts/train_utils.py:196]: Nit: Remove reference of rubik

File: training/distributed_training/pytorch/model_parallel_v2/shared-scripts/train_lib.py (hunk 1 above)
Comment (viclzhu) [reply training/distributed_training/pytorch/model_parallel_v2/shared-scripts/train_lib.py]: Updated!

File: training/distributed_training/pytorch/model_parallel_v2/shared-scripts/train_lib.py (hunk 3 above)
Comment (viclzhu) [reply training/distributed_training/pytorch/model_parallel_v2/shared-scripts/train_lib.py]: Updated!

File: training/distributed_training/pytorch/model_parallel_v2/shared-scripts/train_utils.py (hunk 4 above)
Comment (viclzhu) [reply training/distributed_training/pytorch/model_parallel_v2/shared-scripts/train_utils.py]: Updated!

File: training/distributed_training/pytorch/model_parallel_v2/shared-scripts/train_lib.py (hunk 2 above)
Comment (viclzhu) [reply training/distributed_training/pytorch/model_parallel_v2/shared-scripts/train_lib.py]: Removed all TODO

File: training/distributed_training/pytorch/model_parallel_v2/mixtral/smp-train-mixtral-fsdp-ep-fp8.ipynb [hunk 5]
```diff
@@ -0,0 +1,1043 @@
...
+    "                    \"activation_loading_horizon\": activation_loading_horizon,\n",
+    "                    \"hybrid_shard_degree\": hybrid_shard_degree,\n",
+    "                    \"sm_activation_offloading\": offload_activations,\n",
+    "                    \"expert_parallel_degree\": expert_parallel_degree,\n",
+    "                },\n",
+    "            }\n",
+    "        },\n",
+    "    },\n",
+    "    py_version=\"py310\",\n",
+    "    framework_version=\"2.2.0\",\n",
```
Comment (Teng-xu) [inline training/distributed_training/pytorch/model_parallel_v2/mixtral/smp-train-mixtral-fsdp-ep-fp8.ipynb:850]: framework_version should be 2.3.1

File: training/distributed_training/pytorch/model_parallel_v2/mixtral/smp-train-mixtral-fsdp-ep-fp8.ipynb [hunk 6]
```diff
@@ -0,0 +1,1043 @@
...
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# @title Setup instances: type and count.\n",
+    "\n",
+    "instance_type = \"ml.p5.48xlarge\"\n",
+    "\n",
+    "instance_count = 8\n",
```
Comment (Teng-xu) [inline training/distributed_training/pytorch/model_parallel_v2/mixtral/smp-train-mixtral-fsdp-ep-fp8.ipynb:761]: We don't need 8 p5 nodes to run with this model config, 1 or 2 should be enough

File: training/distributed_training/pytorch/model_parallel_v2/mixtral/smp-train-mixtral-fsdp-ep-fp8.ipynb (hunk 5 above)
Comment (viclzhu) [reply training/distributed_training/pytorch/model_parallel_v2/mixtral/smp-train-mixtral-fsdp-ep-fp8.ipynb]: Updated!

File: training/distributed_training/pytorch/model_parallel_v2/mixtral/smp-train-mixtral-fsdp-ep-fp8.ipynb (hunk 6 above)
Comment (viclzhu) [reply training/distributed_training/pytorch/model_parallel_v2/mixtral/smp-train-mixtral-fsdp-ep-fp8.ipynb]: Updated!

**Comment by viclzhu** [comment]
Closing and re-opening new PR following the new directory structure:
https://github.com/aws/amazon-sagemaker-examples/pull/4739

**Pull request closed without merging**